flow/Tracing.h

```diff
@@ -64,19 +68,36 @@ struct Span {
 		std::swap(parents, other.parents);
 	}
 
-	void addParent(SpanID span) { parents.push_back(arena, span); }
+	void addParent(SpanID span) {
+		if (parents.size() == 0) {
+			// Use first parent to set trace ID. This is non-ideal for spans
+			// with multiple parents, because the trace ID will associate the
+			// span with only one trace. A workaround is to look at the parent
+			// relationships instead of the trace ID. Another option in the
+			// future is to keep a list of trace IDs.
+			context = SpanID(span.first(), context.second());
+		}
+		parents.push_back(arena, span);
+	}
+
+	void addTag(const std::string& key, const std::string value) {
```

Comment:
```suggestion
	void addTag(const std::string& key, const std::string& value) {
```
Why is the value not a reference?

Did you performance test this? I think the key could be a `StringRef` here (like the `Location` type) so this would get rid of one additional allocation. But the value would still needed to be allocated (and copied).

Reply:
Value should be a reference. I'm actually thinking `value` could possibly be a `StringRef` as well. The reason I didn't make these both `StringRef`s is because I thought there was already a copy being made.

When adding a tag, the key will usually be a string literal. When calling `StringRef("test")`, which `StringRef` constructor will be used? My understanding is that the string literal would implicitly have the `std::string` constructor called and would be passed to `StringRef` as a `std::string`.
```
StringRef( const uint8_t* data, int length ) : data(data), length(length) {}
StringRef( const std::string& s ) : data((const uint8_t*)s.c_str()), length((int)s.size()) { ... }
```

The value for a tag may be a `StringRef` itself, but it may also be other types like a `Version` (`int64_t`). In this case, it's necessary to convert the value to a string, requiring `std::to_string`. Converting `std::string` to `StringRef` still copies the pointer, but it could be more efficient than converting `StringRef` values to `std::string` which involves creating a new string.

I have not performance tested this, how would you recommend doing so?

Reply:
Looking at this a bit more and I think both, key and value, should be StringRef. Here's the logic:

- For the key, you can usually use something like `LiteralStringRef("key")` (`StringRef("key")` doesn't work). Copying a string literal is just extra work. If the key is not a literal, the same argument as for the value below applies.
- For values you would push the responsibility to manage the memory to the caller. For spans, this should not be a problem as a span has a very well defined, easy to understand lifetime. I guess in many cases, there's already a `Standalone<StringRef>` that lives as long as the span - if not, you would need to create one (and potentially create a copy, but it would be very explicit).

Reply:
`StringRef("key")` works for me, but passing `"key"` to a function that wants a `StringRef` doesn't work. I think the first is using https://en.cppreference.com/w/cpp/language/converting_constructor. I see how `LiteralStringRef("key")` is better than `StringRef("key")` though, because the first copies a pointer but the second creates a new `std::string` object.